### PR TITLE
[Network] Application Gateway support private link feature

### PR DESCRIPTION
**Pull request by haroldrandom**
**Description**  
Close https://github.com/Azure/azure-cli/issues/13819

**Testing Guide**  

**History Notes**  

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.  
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S172

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 1]
```diff
@@ -148,7 +148,13 @@ def create_application_gateway(cmd, application_gateway_name, resource_group_nam
                                public_ip_address_type=None, subnet_type=None, validate=False,
                                connection_draining_timeout=0, enable_http2=None, min_capacity=None, zones=None,
                                custom_error_pages=None, firewall_policy=None, max_capacity=None,
-                               user_assigned_identity=None):
+                               user_assigned_identity=None,
+                               enable_private_link=False,
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:152]: Could they be updated after the creation? I thought we would have an command group as well.

Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:152]: Added

File: src/azure-cli/azure/cli/command_modules/network/_params.py [hunk 2]
```diff
@@ -146,6 +146,13 @@ def load_arguments(self, _):
         c.argument('enable_http2', arg_type=get_three_state_flag(positive_label='Enabled', negative_label='Disabled'), options_list=['--http2'], help='Use HTTP2 for the application gateway.', min_api='2017-10-01')
         c.ignore('public_ip_address_type', 'frontend_type', 'subnet_type')
 
+    with self.argument_context('network application-gateway', arg_group='Private Link Configuration') as c:
+        c.argument('enable_private_link', action='store_true', help='Enable Private Link feature for this application gateway', default=False)
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/_params.py:150]: should user provide `frontend_ip` as well?

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 3]
```diff
@@ -434,6 +456,275 @@ def show_ag_identity(cmd, resource_group_name, application_gateway_name):
...
+        if subnet.address_prefix == private_link_subnet_prefix:
+            raise CLIError('Subnet prefix duplicates')
+        if subnet.address_prefixes and private_link_subnet_prefix in subnet.address_prefixes:
+            raise CLIError('Subnet prefix duplicates')
+
+    private_link_subnet = Subnet(name=private_link_subnet_name,
+                                 address_prefix=private_link_subnet_prefix,
+                                 private_link_service_network_policies='Disabled')
+    vnet.subnets.append(private_link_subnet)
+    ncf.virtual_networks.create_or_update(resource_group_name, vnet_name, vnet)
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:519]: I think user should be able to create subnet by himself.

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 4]
```diff
@@ -434,6 +456,275 @@ def show_ag_identity(cmd, resource_group_name, application_gateway_name):
...
+            break
+    else:
+        raise CLIError("Priavte Link doesn't exist")
+
+    return target_private_link
+
+
+def list_ag_private_link(cmd,
+                         resource_group_name,
+                         application_gateway_name):
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:575]: Please set this one with `id_part=None` under `list` command

File: src/azure-cli/azure/cli/command_modules/network/custom.py [hunk 5]
```diff
@@ -434,6 +456,275 @@ def show_ag_identity(cmd, resource_group_name, application_gateway_name):
...
+    else:
+        raise CLIError("IP Configuration doesn't exist")
+
+    return target_private_link_ip_config
+
+
+def list_ag_private_link_ip(cmd,
+                            resource_group_name,
+                            application_gateway_name,
+                            private_link_name):
```
Comment (MyronFanQiu) [inline src/azure-cli/azure/cli/command_modules/network/custom.py:684]: set them twoo

Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:684]: fixed

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 4 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:575]: fixed

File: src/azure-cli/azure/cli/command_modules/network/_params.py (hunk 2 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/_params.py:150]: The creation of appgw while specifying frontend IP can take effect only one IP, either the private one or the public one. 
No matter which one is selected, it's named `appGatewayFrontendIP`. 
So, requiring user to specify a --frontend-ip here is unnecessary.
Besides, I left the help message `The frontend IP which the Private Link will associate to` in `--frontend-ip` to indicate user while adding a private link.

File: src/azure-cli/azure/cli/command_modules/network/custom.py (hunk 3 above)
Comment (haroldrandom) [reply src/azure-cli/azure/cli/command_modules/network/custom.py:519]: Will fix in S173

**Pull request merged**